### DISCUSSION TITLE
How to configure interface config post installation

### DISCUSSION THREAD

**Discussion by MohanrajBabu** [cobbler/cobbler#2582] [Q&A]
Hi,

**cobbler 2.8.4**

I am using the sample_end.ks kickstart file to automate centos installation in a virtual machine.  There are two interfaces attached to the VM. 
One is connected to the network to pxe boot (20.1.1.0),
Other is the public network(192.168.43.0), dhcp disabled, so have to configure manually.

Below is the system profile created for the VM. I configured the interface name as  eth0 for the 20.1.1.0 network and eth1 for 192.168.43.0 network.
```
Name                           : Centos-network-test
TFTP Boot Files                : {}
Comment                        : 
Enable gPXE?                   : False
Fetchable Files                : {}
Gateway                        : 192.168.43.1
Hostname                       : Dev-system
Image                          : 
IPv6 Autoconfiguration         : False
IPv6 Default Device            : 
Kernel Options                 : {}
Kernel Options (Post Install)  : {}
Kickstart                      : /var/lib/cobbler/kickstarts/sample_end.ks
Kickstart Metadata             : {}
LDAP Enabled                   : False
LDAP Management Type           : authconfig
Management Classes             : []
Management Parameters          : <<inherit>>
Monit Enabled                  : False
Name Servers                   : ['8.8.8.8']
Name Servers Search Path       : []
Netboot Enabled                : False
Owners                         : <<inherit>>
Power Management Address       : 
Power Management ID            : 
Power Management Password      : 
Power Management Type          : amt_ws
Power Management Username      : 
Profile                        : Centos7-x86_64
Internal proxy                 : <<inherit>>
Red Hat Management Key         : <<inherit>>
Red Hat Management Server      : <<inherit>>
Repos Enabled                  : False
Server Override                : <<inherit>>
Status                         : production
Template Files                 : {}
Virt Auto Boot                 : 0
Virt CPUs                      : <<inherit>>
Virt Disk Driver Type          : <<inherit>>
Virt File Size(GB)             : <<inherit>>
Virt Path                      : <<inherit>>
Virt PXE Boot                  : 0
Virt RAM (MB)                  : <<inherit>>
Virt Type                      : xenpv
Interface =====                : eth1
Bonding Opts                   : 
Bridge Opts                    : 
CNAMES                         : []
InfiniBand Connected Mode      : False
DHCP Tag                       : 
DNS Name                       : 
Per-Interface Gateway          : 
Master Interface               : 
Interface Type                 : 
IP Address                     : 192.168.43.75
IPv6 Address                   : 
IPv6 Default Gateway           : 
IPv6 MTU                       : 
IPv6 Prefix                    : 
IPv6 Secondaries               : []
IPv6 Static Routes             : []
MAC Address                    : 08:00:27:16:56:B0
Management Interface           : False
MTU                            : 
Subnet Mask                    : 255.255.255.0
Static                         : True
Static Routes                  : []
Virt Bridge                    : xenbr0
Interface =====                : eth0
Bonding Opts                   : 
Bridge Opts                    : 
CNAMES                         : []
InfiniBand Connected Mode      : False
DHCP Tag                       : 
DNS Name                       : 
Per-Interface Gateway          : 
Master Interface               : 
Interface Type                 : 
IP Address                     : 20.1.1.151
IPv6 Address                   : 
IPv6 Default Gateway           : 
IPv6 MTU                       : 
IPv6 Prefix                    : 
IPv6 Secondaries               : []
IPv6 Static Routes             : []
MAC Address                    : 08:00:27:2D:94:E1
Management Interface           : False
MTU                            : 
Subnet Mask                    : 255.255.255.0
Static                         : True
Static Routes                  : []
Virt Bridge                    : xenbr0

```

After the installation, I could not able to connect to internet. Also there are four interfaces profiles got created in /etc/sysconfig/network-scripts/ directory as below. Expected only 2 profile in it.

```
[root@Dev-system network-scripts]# ls | grep ifcfg
ifcfg-enp0s3
ifcfg-enp0s8
ifcfg-eth0
ifcfg-eth1
ifcfg-lo
```

ifcfg-enp0s3
```
# Generated by dracut initrd
NAME="enp0s3"
DEVICE="enp0s3"
ONBOOT="yes"
NETBOOT="yes"
UUID="4b328aed-e76b-4ee9-b53d-f9d393f5e42a"
IPV6INIT="yes"
BOOTPROTO="none"
TYPE="Ethernet"
PROXY_METHOD="none"
BROWSER_ONLY="no"
IPADDR="20.1.1.151"
PREFIX="24"
GATEWAY="192.168.43.1"
DNS1="8.8.8.8"
DEFROUTE="yes"
IPV4_FAILURE_FATAL="no"
IPV6_AUTOCONF="yes"
IPV6_DEFROUTE="yes"
IPV6_FAILURE_FATAL="no"
```

ifcfg-eth0
```
DEVICE=eth0
ONBOOT=yes
HWADDR=08:00:27:2D:94:E1
TYPE=Ethernet
BOOTPROTO=none
IPADDR=20.1.1.151
NETMASK=255.255.255.0
DNS1=8.8.8.8
```

ifcfg-enp0s8

```
TYPE="Ethernet"
PROXY_METHOD="none"
BROWSER_ONLY="no"
BOOTPROTO="none"
IPADDR="192.168.43.75"
PREFIX="24"
GATEWAY="192.168.43.1"
DNS1="8.8.8.8"
DEFROUTE="yes"
IPV4_FAILURE_FATAL="no"
IPV6INIT="yes"
IPV6_AUTOCONF="yes"
IPV6_DEFROUTE="yes"
IPV6_FAILURE_FATAL="no"
IPV6_ADDR_GEN_MODE="stable-privacy"
NAME="enp0s8"
UUID="2f238c27-9767-439c-b353-e37f46206fd6"
DEVICE="enp0s8"
ONBOOT="yes"
```

ifcfg-eth1
```
DEVICE=eth1
ONBOOT=yes
HWADDR=08:00:27:16:56:B0
TYPE=Ethernet
BOOTPROTO=none
IPADDR=192.168.43.75
NETMASK=255.255.255.0
DNS1=8.8.8.8
```

network scripts ifcfg-eth0 and ifcfg-eth1 are the device configs.
```
[root@Dev-system network-scripts]# ifdown enp0s3
[root@Dev-system network-scripts]# ifup enp0s3
Error: Connection activation failed: No suitable device found for this connection.
[root@Dev-system network-scripts]# ifdown eth0
[root@Dev-system network-scripts]# ifup eth0
Connection successfully activated (D-Bus active path: /org/freedesktop/NetworkManager/ActiveConnection/7)
[root@Dev-system network-scripts]# 
```
Network 20.1.1.0 is the internal network, it cant connect to internet. Network 192.168.43.0 is the bridged adapter that connect to internet. 
So is it possible in cobbler to set DEFROUTE interface to yes or no.

Why 2 profile created for each interface and how  to fix this issue?

Thanks,
Mohan

**Comment by watologo1** [comment]
On Thursday, October 10, 2019 7:17:49 AM CEST Mohanraj B wrote:
> Hi,
> 
> I am using the sample_end.ks kickstart file to automate centos installation
> in a virtual machine.  There are two interfaces attached to the VM. One is
> connected to the network to pxe boot (20.1.1.0),

This one you certainly want to declare as the "management" interface.
This should be passed via distro specific kernel boot param.
Not sure whether it's implemented for your distro.

Then the installer on know which interface to enable via dhcp
for installation.

> MAC Address                    : 08:00:27:2D:94:E1
> Management Interface           : False

 
> network scripts ifcfg-eth0 and ifcfg-eth1 are the device configs.
> ```
> [root@Dev-system network-scripts]# ifdown enp0s3
> [root@Dev-system network-scripts]# ifup enp0s3
> Error: Connection activation failed: No suitable device found for this
> connection. [root@Dev-system network-scripts]# ifdown eth0
> [root@Dev-system network-scripts]# ifup eth0
> Connection successfully activated (D-Bus active path:
> /org/freedesktop/NetworkManager/ActiveConnection/7) [root@Dev-system
> network-scripts]#
> ```
> Network 20.1.1.0 is the internal network, it cant connect to internet.
But it is enough for complete installation process?

> Network 192.168.43.0 is the bridged adapter that connect to internet. So is
> it possible in cobbler to set DEFROUTE interface to yes or no.
In SUSE this certainly would be autoyast magic.
You have to dig for your distribution how to enable dhcp on both/all network
interfaces by script (and add it as post script to cobbler) or via whatever
auto-install process.

> Why is this problem occurs. How to fix this issue?

The code for what I mentioned first is in (debian and ubuntu seem to
be able to do this):
cobbler/tftpgen.py


        management_interface = None
        if system is not None:
            blended = utils.blender(self.api, False, system)
            # find the first management interface
            try:
                for intf in list(system.interfaces.keys()):
                    if system.interfaces[intf]["management"]:
                        management_interface = intf
                        break


            elif distro.breed == "debian" or distro.breed == "ubuntu":
                append_line = "%s auto-install/enable=true priority=critical netcfg/choose_interface=auto url=%s" % (append_line, autoinstall_path)
                if management_interface:
                    append_line += " netcfg/choose_interface=%s" % management_interface

**Comment by mercury00** [comment]
To partially answer why you're getting four instead of two interfaces; it seems that the centos system (from dracut) identifies two network devices which it automatically creates config files for: enp0s3 and enp0s8.

You've defined two more: eth0 and eth1. 

A quick fix is to rename your eth0 and eth1 interfaces with the system assigned names enp0s3 and enp0s8 (respectively, from what I can tell) - a harder fix would be to interrupt or overwrire the dracut process that is creating the system defined interfaces so it just uses your names instead, eth0 and eth1. (maybe by passing `net.ifnames=0` on the kernel line at boot?)

I'm not familiar enough with that process so I always just name my interfaces the same as the system identifies them with. In recent linux OSes those names are stable (as they identify devices - see https://www.freedesktop.org/wiki/Software/systemd/PredictableNetworkInterfaceNames/ )
